Release/v0.1.0 to master

r4dug opened the pull request.

r4dug committed:
Update README.md

Commit:
Created project structure;
Added logic for finding the closest coffee shops by the user's location;

Commit:
Merge feature/locate_coffee_shops to develop branch;

Commit:
Added feature/unit_tests branch which contains implemented unit tests;

r4dug committed:
Merge pull request #1 from r4dug/feature/unit_tests

Added feature/unit_tests branch which contains implemented unit tests;

Commit:
Added "locate the closest coffee shops" functionality;

r4dug committed:
Merge pull request #2 from r4dug/feature/locate_coffee_shops

Added "locate the closest coffee shops" functionality;

Commit:
Added code explanation;

Commit:
Updated unit tests;

Commit:
Updated unit tests;

alinschuller commented:
Hi Radu! 

Please cleanup the repo (via `.gitignore`), by commiting only meaningful files, there's no need for the `.idea` folder here 😄 

Also, could you please update the README with instructions on how to run this? Thanks a lot!

Commit:
Updated gitignore file;
Added instructions on "how to run the code" in README.md;

Commit:
Updated gitignore file;

Commit:
fixed untracked files

r4dug commented:
Hi Alin! Done. Thanks!

Commit:
Fixed untracked files;

alinschuller commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+# Ignore idea files
+.idea/
```
We should add the venv folder too here, I saw that you had it in a previous commit 😄

r4dug replied:
ye, will do :)

alinschuller commented on src/main.py:
```diff
@@ -0,0 +1,35 @@
+    #return the content of the csv file in a list
+    coffee_shops = list(csv_file)
+    #locate the three closest coffee shops from the user location
+    coffee_shops = locate_closest_coffee_shops(user_loc, coffee_shops)
```
Unfortunately, this fails with the following error:
`TypeError: locate_closest_coffee_shops() missing 1 required positional argument: 'coffee_shops'`

r4dug replied:
oups, should have been user's input as X and Y coordinates instead of user_loc

alinschuller commented on src/locate_coffee_shop.py:
```diff
@@ -0,0 +1,35 @@
+    for shop in coffee_shops:
+        try:
+            # find distance between two points
+            shop_distance = get_distance_between_two_points(user_x, user_y, shop[1], shop[2])
```
I see that we access the file content via indices (shop[0], shop[1], shop[2]).

Can you suggest another way to access them to improve readability and be less prone to error?

r4dug replied:
Yes, one possibility could have been using list comprehension and the enumerate() function.

alinschuller commented on src/locate_coffee_shop.py:
```diff
@@ -0,0 +1,35 @@
+import math
+from collections import namedtuple
+
+ProximateCoffeeShop = namedtuple("CoffeeShop", "name distance")
```
Nice, finally someone that used named tuples

r4dug replied:
😉

alinschuller commented on src/main.py:
```diff
@@ -0,0 +1,35 @@
+    user_loc = [args.user_x, args.user_y]
+    #use urllib3 in order to get the http request of the provided URL
+    http = urllib3.PoolManager()
+    response = http.request('GET', args.shops_url)
```
This is definitely a good approach to read a csv file, there are also the `requests` and `pandas` libraries, each providing some benefits.

Which would do you think would be best in this case?

r4dug replied:
Hmm.. on second thought my opinion is requests could have been a good choice as well (this was my first approach).
Now I would go with something like this:

import requests
from contextlib import closing
import csv
from codecs import iterdecode

url = ""

with closing(requests.get(url, stream=True)) as r:
    reader = iterdecode(csv.reader(r.iter_lines(), 'utf-8'), 
                        delimiter=',', 
                        quotechar='"')
    for row in reader:
        print(row)

*By setting stream=True in the GET request, when we pass r.iter_lines() to csv.reader(), we are passing a generator to csv.reader(). By doing so, we enable csv.reader() to lazily iterate over each line in the response with for row in reader.

This avoids loading the entire file into memory before we start processing it, drastically reducing memory overhead for large files.

r4dug replied:
I remember at some point using 'pandas' with 'read chunks' for something similar, but I was saving data in a dataframe.

alinschuller replied:
Pandas can simplify a lot your code, next to the fact that it can read both local and url files.

So you would just have a read like:
`csv_file = pd.read_csv(args.shop_url, header=None, names=["coffee_shop_name", "x_coordinate", "y_coordinate"])`
This way, next to the fact that it knows that `x_coordinate` is a float, you would be able to iterate each coffee shop and access it like `shop.coffee_shop_name`

r4dug replied:
Yes, agree. Thanks.

alinschuller replied:
@r4dug was curious about your own way with `requests` library, not this one https://stackoverflow.com/a/38677619 😞

r4dug replied:
@alinschuller I understand your point of view, you are right. I copied the text from 'stack' as I found it well explained, it doesn't mean I am not aware of the differences between those libraries. But again, I totally understand the fact you wanted to hear my own story.

alinschuller commented on tests/unit_test_coffee_shop.py:
```diff
@@ -0,0 +1,51 @@
+class TestLocateCoffeeShop(unittest.TestCase):
+
+    # test with a wrong file path
+    def test_locate_coffee_shop_no_file(self):
```
This fails, instead of the `input file does not exist` error I get `string index out of range`.

Also, this is not the best approach for checking exceptions, but by using the `assertRaises` method. 
Is there a reason why you didn't go with that?

r4dug replied:
Yes, I am aware assertRaises method is a better choice. To be honest, I forgot to delete this test which I didn't intend to keep.

alinschuller replied:
Why not? The test case itself is good, we should be testing what happens when the file is not provided as expected

r4dug replied:
As test cases are concerned, I couldn't agree more to the fact that plenty of them could have been added. The 'input file does not exist' test case is fine, but I am not happy with the way it was written.

alinschuller commented on tests/unit_test_coffee_shop.py:
```diff
@@ -0,0 +1,51 @@
+        except Exception as ex:
+            assert "Input file does not exist." in str(ex)
+
+    def test_locate_coffee_shop(self):
```
This test fails as well, main reason being the fact that the distance is not being rounded up.

So for example, instead of having `1.4142` for `Loc 1`, you get `1.4142135623730951`

r4dug replied:
Yes, I was aware rounding up the value will fix this.

alinschuller commented on tests/unit_test_coffee_shop.py:
```diff
@@ -0,0 +1,51 @@
+        self.assertEqual(locate_closest_coffee_shops(user_location[0], user_location[1], coffee_shops),
+                         expected_response)
+
+    def test_locate_coffee_shop_fail(self):
```
This throws the following error:
`unsupported operand type(s) for -: 'int' and 'str`.

It definitely makes sense why based on what's given in `coffee_shops`. I would rather create separate test cases. So I would assert that an error is raised for the above.

r4dug replied:
👍

alinschuller commented on src/main.py:
```diff
@@ -0,0 +1,35 @@
+        print(f"The URL you provided generated a code {response.status}: {response.reason}")
+        exit()
+    #read the csv file found at the particular location
+    csv_file = csv.reader(response.data.decode('utf-8').split("\n"))
```
Aren't all the coordinates stored as strings? If so, they would fail because there is no conversion to float values.

r4dug replied:
Yes, definitely coordinates should have been converted to float values.

alinschuller reviewed (commented):
Overall, I like a lot of things from this implementation, you make good use of existing libraries and the code is simple and readable.

I do have a couple things though that I would’ve liked to see but also a couple issues:

1. I would do more validations, we do not want a user to pass a string and be hit up with a float validation, they most likely don’t know what that means or pass a string to the url and be hit up with `nodename nor servname provided`.

2. The solution cannot be executed, it fails. The only way I am able to run it, is by updating the code myself.

3. Tests fail/throw errors. 
- Tests are meant to ensure that our code is working as expected, including adding test cases for error handling. 
- If all our tests fail/throw errors, in a CI pipeline, CI simply won’t pass.
- Testing `main` should’ve pointed out the issues with the implementation.


Could you detail your experience when executing this solution? Have you encountered the same issues?

Thanks a lot and have a great day!

r4dug commented:
Hey, @alinschuller ! I've updated the code with the following:
1) locate_closest_coffee_shops function in locate_coffee_shop.py has been simplified/corrected based on your suggestions;
2) main.py file has beed updated and now 'requests' library is being used for getting the http request, parsing the csv file at the particular location and saving the closest three coffee shops as per user's input coordinates;
3) venv has been included in .gitignore
*code runs now without any issues:
Command I used: python3 main.py 45.123 -123.4567 https://raw.githubusercontent.com/Agilefreaks/test_oop/master/coffee_shops.csv

Commit:
1)locate_closest_coffee_shops function in locate_coffee_shop.py has been simplified/corrected;
2) main.py file has beed updated and now 'requests' library is being used instead of urllib3;
3) venv was added in .gitignore;

Commit:
Reformat code;

Commit:
Removed wrong unit test;

alinschuller commented:
@r4dug thanks a lot for looking over my comments and providing quick feedback/changes to them.

Now the solution is definitely working. We'll be following up soon.

Have a great day!
